### PR TITLE
Perf: Speed up CPU and GPU NMF by 30% or more

### PR DESCRIPTION
**Pull request by jberg5**
Speedup comes from a few things:
- Removing the ones() matrix allocation
- Skipping the two `ones @ ht` and `wt @ ones` matmuls
- One-dimensional case, which is always true for cpu and the default for the gpu, was paying the torch batched matrix multiply kernel overhead. It's not huge, but shows up for tight iterative loops like we're doing here.

This ends up being much closer to what [scikit-learn's NMF implementation](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/decomposition/_nmf.py) does.

On my macbook, speedups look like:

```
  ┌──────┬───────┬───────────┬───────────────┬─────────┐                                                                                        
  │ rank │ iters │ master ms │ faster-nmf ms │ speedup │                                                                                        
  ├──────┼───────┼───────────┼───────────────┼─────────┤                                                                                        
  │    3 │  1000 │      55.0 │          32.8 │   1.68× │                                                                                        
  ├──────┼───────┼───────────┼───────────────┼─────────┤                                                                                        
  │    5 │  1000 │      50.1 │          27.1 │   1.85× │                                                                                        
  ├──────┼───────┼───────────┼───────────────┼─────────┤
  │    5 │  5000 │     234.9 │         129.5 │   1.81× │                                                                                        
  ├──────┼───────┼───────────┼───────────────┼─────────┤                                                                                        
  │   10 │  1000 │      54.4 │          27.8 │   1.96× │
  ├──────┼───────┼───────────┼───────────────┼─────────┤                                                                                        
  │   10 │  5000 │     261.3 │         149.8 │   1.74× │ 
  └──────┴───────┴───────────┴───────────────┴─────────┘    
```
 
running https://gist.github.com/jberg5/c2d59045f97dce18d70615993986f2d3 on my branch and master and comparing results. (full transparency: claude wrote that benchmark!). Results are equal to f32 tolerance - float reduction order is changed so I wouldn't expect bitwise identical outputs.

I also ran sigProfilerExtractor on the Nik-Zainal 2012 21-BRCA catalogue with `nmf_replicates=100`, `minimum_signatures=1, maximum_signatures=10`, and `max_nmf_iterations=100000` (all other options used the defaults). Used a gcloud c3-standard-8 (Intel Xeon Platinum 8481C, Sapphire Rapids, 4 physical cores, OMP/MKL/OPENBLAS/NUMEXPR_NUM_THREADS=1).

I ran this master twice and this branch twice, ABAB, and all four runs selected the same rank (3), with identical reported stability (0.98), identical Frobenius% (5.78%), and identical sample-level  statistics. This branch was consistently 20 minutes faster (72 minutes -> 52 minutes).

**GPU Performance**: I spun up an ec2 with a T4 to do some benchmarking there and to verify correctness:

```
    ┌──────┬───────┬──────┬───────────┬───────────────┬─────────┐                                                                               
    │   M  │   N   │ rank │ master ms │ faster-nmf ms │ speedup │                                                                               
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤                                                                               
    │   96 │    21 │   5  │     405.3 │         307.5 │  1.32×  │                                                                               
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤                                                                               
    │   96 │    21 │  10  │     407.4 │         308.4 │  1.32×  │                                                                               
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤                                                                               
    │   96 │   500 │   5  │     418.8 │         314.5 │  1.33×  │                                                                               
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤
    │   96 │   500 │  10  │     417.9 │         315.4 │  1.32×  │                                                                               
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤                                                                               
    │   96 │  5000 │   5  │     419.8 │         318.4 │  1.32×  │                                                                               
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤                                                                               
    │   96 │  5000 │  10  │     420.9 │         319.1 │  1.32×  │
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤                                                                               
    │   96 │ 10000 │  10  │     555.8 │         461.8 │  1.20×  │
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤                                                                               
    │ 1536 │   500 │  10  │     447.6 │         387.6 │  1.15×  │                                                                               
    ├──────┼───────┼──────┼───────────┼───────────────┼─────────┤                                                                               
    │ 1536 │  5000 │  10  │    4130.2 │        3195.2 │  1.29×  │                                                                               
    └──────┴───────┴──────┴───────────┴───────────────┴─────────┘
```

### CODE REVIEW

**Comment by mishugeb** [comment]
Hello, 
Thanks for the pull request. I am sorry that we cannot merge it unless it includes a comprehensive accuracy benchmark result using the datasets described at https://www.cell.com/cell-genomics/fulltext/S2666-979X(22)00124-0?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2666979X22001240%3Fshowall%3Dtrue

**Comment by jberg5** [comment]
@mishugeb thanks for taking a look. Can you point me at any other merged PRs that ran this accuracy benchmark, so I can make sure I'm running the specific workflow you're looking for?

For this PR, I confirmed that the results are equivalent (within floating point tolerance). Outputs and summary stats are identical.
